src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -231,21 +255,49 @@ private void DisposeAdaptedConnections()
 
         public void Tick(DateTimeOffset now)
         {
-            Debug.Assert(_frame != null, $"nameof({_frame}) is null");
+            Debug.Assert(_frame != null, $"{nameof(_frame)} is null");
 
             var timestamp = now.Ticks;
 
-            // TODO: Use PlatformApis.VolatileRead equivalent again
-            if (timestamp > Interlocked.Read(ref _timeoutTimestamp))
+            if (timestamp > Interlocked.Read(ref _timeoutTimestamp)) // TODO: Use PlatformApis.VolatileRead equivalent again
             {
                 CancelTimeout();
 
                 if (_timeoutAction == TimeoutAction.SendTimeoutResponse)
                 {
+                    SetTimeoutResponse();
+                }
+
+                Timeout();
+            }
+            else if (_timingReads)
+            {
+                _readTimingElapsed += timestamp - _lastTimestamp;
+
+                if (_readTimingElapsed > _frame.RequestBodyTimeout.Ticks)
+                {
+                    Log.RequestBodyTimeout(_context.ConnectionId, _frame.TraceIdentifier, _frame.RequestBodyTimeout);
                     Timeout();
                 }
+                else if (_frame.RequestBodyMinimumDataRate > 0)
+                {
+                    var rate = (_bytesRead / (_readTimingElapsed / TimeSpan.TicksPerSecond));
```

Comment:
I would clarify this line of code by splitting it into a couple lines, ex:

```
var elapsedSecondsFloored = _readTimingElapsed / TimeSpan.TicksPerSecond;
var rate = _bytesRead / elapsedSecondsFloored ;
```

Technically the rate is floored too which isn't exactly ideal, but bytes/second is a precise enough measurement that I doubt it matters.